File: hackerrank-scala/src/algorithms/sort/Quicksort.scala
package algorithms.sort

import scala.io.StdIn
import scala.util.Random

object QuickSort {
  
   def main(args: Array[String]){
     
     val length = StdIn.readInt()
     var array =  StdIn.readLine().split(" ").map(_.toInt)
     quickSort(array)
     print(array, 0, array.size)
   }
   
   private def quickSort(array: Array[Int]) {
     var vars = Random.shuffle(array.toList).toArray
     sort(array, 0, array.size - 1)
   }
   
   // will partition | <= j | j | >= j |
   // returns j - the index of the partition item
   // takes the 1st value as the initial hook
   private def partition(array: Array[Int], lo: Int, hi: Int) : Int = {
     var i = lo + 1
     var j = hi
     while(i < j){
       // go right as long as the value found is less than the 1st value (pivot)
       while(array(i) < array(lo) && i < hi) i += 1
       // go left as long as the value found is bigger than the 1st value (pivot)
       while(array(lo) < array(j) && j >= lo) j -= 1
       if(i < j) swap(array, i, j) 
     }
     if(array(lo) > array(j)) swap(array, lo, j)
     j
   }
   
   private def swap(array: Array[Int], i: Int, j: Int) {
     val swap = array(i)
     array(i) = array(j)
     array(j) = swap
   }
   
   private def sort(array: Array[Int], lo: Int, hi: Int) {
     if(hi <= lo) return
     val j = partition(array, lo, hi)
     sort(array, lo, j-1)
     sort(array, j+1, hi)
   }
   
   private def print(array: Array[Int], from: Int, to: Int) {
     val array2 = array.slice(from, to)
     array2.foreach(v => printf("%.0f ".format(v.toDouble)))
     println()
   }
}

File: creek_java/src/io/creek/QuickSort.java
package io.creek;

import java.util.Comparator;

public class QuickSort {

	public static <T> void quickSort(T[] array, Comparator<T> comparator) {
		sort(array, comparator, 0, array.length - 1);
	}

	private static <T> void sort(T[] array, Comparator<T> comparator, int lo, int hi) {
		if (lo >= hi) return;
		int partition = partition(array, comparator, lo, hi);
		sort(array, comparator, lo, partition - 1);
		sort(array, comparator, partition + 1, hi);
	}

	private static <T> int partition(T[] array, Comparator<T> comparator, int lo, int hi) {
		int pivot = lo, left = lo + 1, right = hi;
		while (true) {
			while (left < hi && comparator.compare(array[left], array[pivot]) <= 0) left ++;
			while (right > lo && comparator.compare(array[right], array[pivot]) >= 0) right --;
			if (left >= right) break;
			swap(array, left, right);
		}
		swap(array, lo, right);
		return right;
	}
	
	private static <T> void swap(T[] array, int i, int j) {
		T tmp = array[i];
		array[i] = array[j];
		array[j] = tmp;
	}
}
